vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go

```diff
@@ -0,0 +1,166 @@
+		return nil
+	}
+	if attr.GetKind().GroupKind() != coreapi.Kind("Pod") {
+		return nil
+	}
+	if len(attr.GetSubresource()) != 0 {
+		return nil
+	}
+	pod, ok := attr.GetObject().(*coreapi.Pod)
+	if !ok {
+		return admission.NewForbidden(attr, fmt.Errorf("unexpected Type %T", attr.GetObject()))
+	}
+	return p.validatePodTolerations(attr, pod.Spec)
+}
+
+func (p *podTolerations) validatePodTolerations(attr admission.Attributes, podSpec coreapi.PodSpec) error {
+	// if a pod has no tolerations, then no need to validate it's tolerations
+	if len(podSpec.Tolerations) == 0 {
+		return nil
+	}
+	tolerations := podSpec.Tolerations
+	if len(podSpec.ServiceAccountName) == 0 {
+		return admission.NewForbidden(attr, fmt.Errorf("serviceaccounts are required on pods that want to use tolerations"))
+	}
+	for _, toleration := range tolerations {
+		allow := p.IsAuthorized(podSpec.ServiceAccountName, attr, toleration)
+		var errs []error
+		if !allow {
+			errs = append(errs, fmt.Errorf("serviceaccounts doesn't have permission to provide the given toleration %v on the pod", toleration.Key))
+		}
+		if len(errs) > 0 {
+			err := v1helpers.NewMultiLineAggregate(errs)
+			return admission.NewForbidden(attr, err)
+		}
+	}
+	return nil
+}
+
+func (p *podTolerations) IsAuthorized(podSAName string, attr admission.Attributes, toleration coreapi.Toleration) bool {
+	saInfo := serviceaccount.UserInfo(attr.GetNamespace(), podSAName, "")
+	// check for hard-coded serviceaccount groups
+	SAInNamespace := serviceaccount.UserInfo(attr.GetNamespace(), "system:serviceaccounts:"+attr.GetNamespace(), "")
+	SAAcrossAllNamespace := serviceaccount.UserInfo("", "system:serviceaccounts", "")
+	return p.checkPodTolerationAccess(saInfo, attr, toleration) || p.checkPodTolerationAccess(SAInNamespace, attr, toleration) || p.checkPodTolerationAccess(SAAcrossAllNamespace, attr, toleration)
+}
+
+// buildTolerationsAuthorizationAttributes builds the tolerations attributes from the user
+func buildTolerationsAuthorizationAttributes(info user.Info, namespace string, toleration coreapi.Toleration) authorizer.Attributes {
+	verb := toleration.Operator
+	tolerationKey := toleration.Key
+	if len(toleration.Key) == 0 {
+		tolerationKey = masterToleration
+	}
+	authzAttr := authorizer.AttributesRecord{
+		User:            info,
+		Verb:            string(verb),
+		Namespace:       namespace,
+		Resource:        tolerationKey,
+		APIGroup:        "toleration.scheduling.openshift.io",
+		ResourceRequest: true,
+		Name:            string(toleration.Effect) + ":" + toleration.Value,
+	}
+	return authzAttr
+}
+
+// checkPodTolerationAccess checks if pods are allowed to have a particular toleration. In this case, we're focusing
+// on master tolerations. For example,
+/*
+tolerations:
+- effect: NoSchedule
+  key: node-role.kubernetes.io/master
+  operator: Exists
+
+are converted to
+
+apiGroup: toleration.scheduling.openshift.io
+Resource: node-role.kubernetes.io/master
```

Comment:
Is there a way to wildcard by label namespace?  There will be certain label sets that nodes shouldn't be allowed to label.  I suspect we're going to end up needing to have at minimum by namespace, by prefix, etc.

Can we use `namespace` in the rule for `node-role.kubernetes.io` or are you leveraging namespace?  I.e. then the label "name" == `resourceName`, the label namespace == `namespace`, and the resource == `tolerations`?

Reply:
Using the namespace as a real namespace means that an admin granted powers in one namespace can delegate those power to multiple service accounts.  If we override the meaning of namespace, then someone with the power to grant rbac roles in the those namespaces becomes unexpectedly  powerful.

How many different toleration keys do we really expect to have?

Reply:
Limiting based on namespace is really problematic. @smarterclayton, is your concern more on the lines of we will end up having writing permissions for too many tolerations on the pod spec(across multiple repos)?